gateway/ln-gateway/src/state_machine/complete.rs

```diff
@@ -174,28 +177,40 @@ impl CompleteHtlcState {
         common: GatewayCompleteCommon,
         outcome: HtlcOutcome,
     ) -> Result<(), CompleteHtlcError> {
-        let htlc = match outcome {
-            HtlcOutcome::Success(preimage) => InterceptHtlcResponse {
-                action: Some(Action::Settle(Settle {
-                    preimage: preimage.0.to_vec(),
-                })),
-                incoming_chan_id: common.incoming_chan_id,
-                htlc_id: common.htlc_id,
-            },
-            HtlcOutcome::Failure(reason) => InterceptHtlcResponse {
-                action: Some(Action::Cancel(Cancel { reason })),
-                incoming_chan_id: common.incoming_chan_id,
-                htlc_id: common.htlc_id,
-            },
-        };
-
-        // TODO: Can we retry this instead of failing?
-        context
-            .lnrpc
-            .complete_htlc(htlc)
-            .await
-            .map_err(|_| CompleteHtlcError::FailedToCompleteHtlc)?;
-        Ok(())
+        // Wait until the lightning node is online to complete the HTLC
+        loop {
+            let htlc_outcome = outcome.clone();
+            let lightning_context = context.gateway.get_lightning_context().await;
+            match lightning_context {
+                Ok(lightning_context) => {
+                    let htlc = match htlc_outcome {
+                        HtlcOutcome::Success(preimage) => InterceptHtlcResponse {
+                            action: Some(Action::Settle(Settle {
+                                preimage: preimage.0.to_vec(),
+                            })),
+                            incoming_chan_id: common.incoming_chan_id,
+                            htlc_id: common.htlc_id,
+                        },
+                        HtlcOutcome::Failure(reason) => InterceptHtlcResponse {
+                            action: Some(Action::Cancel(Cancel { reason })),
+                            incoming_chan_id: common.incoming_chan_id,
+                            htlc_id: common.htlc_id,
+                        },
+                    };
+
+                    lightning_context
+                        .lnrpc
+                        .complete_htlc(htlc)
+                        .await
+                        .map_err(|_| CompleteHtlcError::FailedToCompleteHtlc)?;
+                    return Ok(());
+                }
+                Err(_) => {
+                    warn!("Trying to complete HTLC but the lightning node is not connected");
```

Comment:
In fact you shouldn't make the assumption the error is about disconnection. Just make it generic:
```suggestion
                Err(e) => {
                    warn!("Trying to complete HTLC but got {e}, will keep retrying...");
```

Reply:
Done